napari/components/_layer_slicer.py

```diff
@@ -43,9 +46,68 @@ def __init__(self):
         """
         self.events = EmitterGroup(source=self, ready=Event)
         self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._force_sync = False
         self._layers_to_task: Dict[Tuple[Layer], Future] = {}
         self._lock_layers_to_task = RLock()
 
+    @contextmanager
+    def force_sync(self):
+        """Context manager to allow a temporary forced synchronous slice.
+        This method only holds the _force_sync variable as True while the
+        manager is open, then resets it back to False after the manager is
+        closed.
+        """
+        self._force_sync = True
+        yield None
+        self._force_sync = False
+
+    def await_slice(self, future: Future, timeout: float = 5) -> None:
+        """Wait for a single slicing task to complete
+
+        Attributes
+        ----------
+        future: concurrent.futures.Future
+            asynchronous slicing task
+        timeout: float
+            time in seconds to wait before raising TimeoutError, defaults to 5s
+
+        Raises
+        ------
+        TimeoutError: when the timeout limit has been exceeded and the task is
+            not yet complete
+        """
+        done, _ = wait([future], timeout=timeout)
+        if done:
+            return
+        else:
+            raise TimeoutError(
+                f'Slice task did not complete within timeout ({timeout}s).'
+            )
+
+    def wait_until_idle(self, timeout: Optional(float or None) = None) -> None:
```

Comment:
This method is more useful than `await_slice` because it uses some private state (i.e. `_layers_to_task`), but I'm still not sure if we should expose it yet unless we can think of an immediate use for it in napari or in a napari plugin.

One idea was that this could be part of a way to make `Viewer.screenshot` to wait for all slicing to be done before capturing a screenshot of the canvas (e.g. wait for all slicing tasks to be done, block any new ones, process all events on the event loop, then finally capture a screenshot). But maybe it should just be added then?

As `LayerSlicer` itself is private, I don't mind adding this now. But if it ever becomes public, we may want to re-evaluate.

Reply:
I can't say I put a lot of thought into making a public function vs private. I'm happy to just make it private. 

I do think it provides a useful purpose (screenshotting at the very least). I expect it would be called by `Viewer.screenshot` once we make it to that level of implementation. 

How do you feel about just making it private and leaving it in?

Reply:
Since `LayerSlicer` has a private name I think it's fine to leave it as public right now. That way it's clear this is intended to be part of the public API of this class (i.e. this is not an implementation detail or helper function).